posthog/api/survey.py

```diff
@@ -0,0 +1,163 @@
+            "targeting_flag",
+            "targeting_flag_filters",
+            "questions",
+            "conditions",
+            "appearance",
+            "created_at",
+            "created_by",
+            "start_date",
+            "end_date",
+        ]
+        read_only_fields = ["id", "linked_flag", "targeting_flag", "created_at"]
+
+    def validate(self, data):
+        linked_flag_id = data.get("linked_flag_id", None)
+        if linked_flag_id:
+            try:
+                FeatureFlag.objects.get(pk=linked_flag_id)
+            except FeatureFlag.DoesNotExist:
+                raise serializers.ValidationError("Feature Flag with this ID does not exist")
+
+        return data
+
+    def create(self, validated_data):
+        validated_data["team_id"] = self.context["team_id"]
+
+        if validated_data.get("targeting_flag_filters", None):
+            # create a new feature flag using the targeting flag filters
+            targeting_flag_filters = validated_data["targeting_flag_filters"]
+            feature_flag_key = slugify(f"survey-targeting-{validated_data['name']}")
+            feature_flag_serializer = FeatureFlagSerializer(
+                data={
+                    "key": feature_flag_key,
+                    "name": f"Targeting flag for survey {validated_data['name']}",
+                    "filters": targeting_flag_filters,
+                },
+                context=self.context,
+            )
+
+            feature_flag_serializer.is_valid(raise_exception=True)
+            targeting_feature_flag = feature_flag_serializer.save()
+            validated_data["targeting_flag_id"] = targeting_feature_flag.id
+            validated_data.pop("targeting_flag_filters")
+
+        validated_data["created_by"] = self.context["request"].user
+        survey: Survey = super().create(validated_data)
+        return survey
+
+    def update(self, instance: Survey, validated_data):
+        # if the target flag filters come back with data, update the targeting feature flag if there is one, otherwise create a new one
+        if validated_data.get("targeting_flag_filters", None):
+            if instance.targeting_flag:
+                existing_targeting_flag = FeatureFlag.objects.get(pk=instance.targeting_flag.id)
+                serialized_data_filters = {
+                    **existing_targeting_flag.filters,
+                    **validated_data["targeting_flag_filters"],
+                }
+                existing_flag_serializer = FeatureFlagSerializer(
+                    existing_targeting_flag,
+                    data={"filters": serialized_data_filters},
+                    partial=True,
+                    context=self.context,
+                )
+                existing_flag_serializer.is_valid(raise_exception=True)
+                existing_flag_serializer.save()
+            else:
+                targeting_flag_key = slugify(f"survey-targeting-{instance.name}")
```

Comment:
I'd recommend pulling this out into a function, which will prevent future bugs (updating it in create but not update)